llvm/lib/IR/RepoTicket.cpp

```diff
@@ -78,186 +87,236 @@ const Constant *getAliasee(const GlobalAlias *GA) {
   for (const GlobalObject *const G : Contributions) {
     GOHash.update(GOIMap[G].InitialDigest.Bytes);
   }
-  return !Contributions.empty();
+  // GOHash/Module is changed if Contributions is not empty.
+  Changed = Changed || !Contributions.empty();
 }
 
-// Loop through the dependences and add the final digest of each global object
-// inside of the dependences to GOHash.
 template <typename Function>
-static bool updateDigestUseDependencies(MD5 &GOHash, const GOVec &Dependencies,
-                                        GOStateMap &Visited, GOInfoMap &GOIMap,
-                                        Function AccumulateGODigest) {
-  bool Changed = false;
+auto HashCache::updateDigestUseDependencies(const GlobalObject *GO, MD5 &GOHash,
+                                            unsigned GOVisitedIndex,
+                                            const GOVec &Dependencies,
+                                            Function AccumulateGODigest)
+    -> std::tuple<size_t, DigestType> {
+  auto LoopPoint = std::numeric_limits<size_t>::max();
   for (const GlobalObject *const G : Dependencies) {
     const llvm::Function *const Fn = dyn_cast<const llvm::Function>(G);
     // if function will not be inlined and not be discarded if it is not used,
     // skip it.
     if (Fn && Fn->hasFnAttribute(Attribute::NoInline) &&
         !Fn->isDiscardableIfUnused())
       continue;
-    Changed = updateDigestUseDependenciesAndContributions(
-                  G, GOHash, Visited, GOIMap, AccumulateGODigest) ||
-              Changed;
+    size_t GVisitedIndex;
+    DigestType GDigest;
+    std::tie(GVisitedIndex, GDigest) =
+        updateDigestUseDependenciesAndContributions<Function>(
+            G, AccumulateGODigest);
+    // A GO which loops back to itself doesn't count as a loop.
+    if (G != GO)
+      LoopPoint = std::min(LoopPoint, GVisitedIndex);
+    GOHash.update(GDigest.Bytes);
   }
-  return Changed;
+  // Encoded the 'End' in the GOhash after accumulating all GO's Dependencies.
+  GOHash.update(static_cast<char>(Tags::End));
+
+  DigestType Digest;
+  GOHash.final(Digest);
+
+  // If LoopPoint is bigger than GOVisitedIndex, GO doesn't belong to any loops
+  // (except self-loop). Record GO's hash in memoized hashes.
+  if (LoopPoint > GOVisitedIndex) {
+    GOHashCache[GO] = Digest;
+    LLVM_DEBUG(dbgs() << "Recording result for \"" << GO->getName() << "\"\n");
+    ++NumMemoizedHashes;
+    ++VisitedCachedGOTimes;
+  } else {
+    ++VisitedUnCachedGOTimes;
+  }
+  return std::make_tuple(LoopPoint, Digest);
 }
 
-// Update the GO's hash value by adding the hash of its dependencies and
-// contributions.
 template <typename Function>
-static bool updateDigestUseDependenciesAndContributions(
-    const GlobalObject *GO, MD5 &GOHash, GOStateMap &Visited, GOInfoMap &GOIMap,
-    Function AccumulateGODigest) {
-  if (GO->isDeclaration())
-    return false;
+auto HashCache::updateDigestUseDependenciesAndContributions(
```

Comment:
Is templating on the function actually necessary (e.g. for performance)?

This PR doesn't introduce it but it does spread it from 1 function (updateDigestUseDependenciesAndContributions) to 3 (updateDigestUseDependencies, updateDigestUseDependenciesAndContributions and calculateDigest) so I think it's worth considering whether it is a good idea.

In trying to understand this code I sketched how it would be if it was replaced with a boolean parameter to select between the two accumulate behaviours instead of passing a function around (https://github.com/SNSystems/llvm-project-prepo/commits/replace_template_with_parameter). I find this easier to understand.

Reply:
Good point. Changed the implementation following your idea.